### PR TITLE
BUGFIX: edgecasy parenthesis in expressions

### PR DESCRIPTION
**Pull request by mhsdesign**
previously `ExpressionNode::fromString("(foo)")` couldnt be parsed (it worked though when used in return statement of a module)

but the real fix, is that you couldnt write `return ((foo))` as $tokens is reassigned but not passed by ref

`(foo === ((null))) ? 1 : (((0)))` this seemed to worked previously but not this: `(((foo)) === ((null))) ? 1 : (((0)))`

### CODE REVIEW

File: src/Parser/Tokenizer/Scanner.php [hunk 1]
```diff
@@ -165,4 +165,26 @@ public static function isEnd(\Iterator $tokens): bool
...
+    /**
+     * @param \Iterator<mixed,Token> $tokens
+     */
+    public static function debugPrint(\Iterator &$tokens): string
+    {
+        $tokens = (function(): \Generator {
+            throw new \Exception('Once debugged, $tokens is empty.');
+            // @phpstan-ignore-next-line
+            yield;
+        })();
```
Comment (grebaldi) [inline src/Parser/Tokenizer/Scanner.php:178]: This should rather be removed, right?

File: src/Parser/Tokenizer/Tokenizer.php [hunk 2]
```diff
@@ -44,7 +44,10 @@ public static function fromSource(Source $source): Tokenizer
      */
     public function getIterator(): \Iterator
     {
-        yield from self::block($this->source->getIterator());
+        $fragments = $this->source->getIterator();
+        while ($fragments->valid()) {
+            yield from self::block($fragments);
+        }
```
Comment (grebaldi) [inline src/Parser/Tokenizer/Tokenizer.php:50]: `while ($fragments->valid())` is imho tautological to `yield from`. At least, when I revert these lines, tests stay green.

What was your reasoning for putting the source iterator into a variable?

File: src/Parser/Tokenizer/LookAhead.php [hunk 3]
```diff
@@ -68,8 +70,11 @@ public function shift(): void
...
     }
 
-    public function type(): TokenType
+    public function type(): ?TokenType
     {
+        if (Scanner::isEnd($this->tokens)) {
+            return null;
+        }
         return Scanner::type($this->tokens);
     }
```
Comment (grebaldi) [inline src/Parser/Tokenizer/LookAhead.php:79]: Why is this needed? (Here also: When I revert those lines, tests stay green)

File: src/Parser/Tokenizer/LookAhead.php [hunk 4]
```diff
@@ -58,7 +58,9 @@ public function getIterator(): \Iterator
             yield $token;
         }
 
-        yield from $this->tokens;
+        if (!Scanner::isEnd($this->tokens)) {
+            yield from $this->tokens;
+        }
```
Comment (grebaldi) [inline src/Parser/Tokenizer/LookAhead.php:63]:
```suggestion
        yield from $this->tokens;
```

`Scanner::isEnd` only checks whether `$this->tokens` is still valid. This is implied in `yield from` - the extra check is not needed.

File: src/Parser/Ast/ExpressionNode.php [hunk 5]
```diff
@@ -51,7 +52,7 @@ public static function fromString(string $expressionAsString): self
      * @param Precedence $precedence
      * @return self
      */
-    public static function fromTokens(\Iterator $tokens, Precedence $precedence = Precedence::SEQUENCE): self
+    public static function fromTokens(\Iterator &$tokens, Precedence $precedence = Precedence::SEQUENCE): self
```
Comment (grebaldi) [inline src/Parser/Ast/ExpressionNode.php:55]: Why pass by reference? Here also: If I revert this, tests stay green.

**Review by grebaldi — COMMENTED** [review]
Hi @mhsdesign,

sorry that I'm taking so long to review your changes. Please know, that I really appreciate the hard work you're putting into this. I only have trouble keeping up with your pace :D

For this particular fix, I think I understand the problem, you're trying to solve. Yet, there's a couple of questions that arose while I was reading the code. But I think, it'll be easy to clear those up and get this fix on the road :)

**EDIT:** I may stand corrected on my comments below, because as pointed out in https://github.com/PackageFactory/PackageFactory.ComponentEngine/pull/10#issuecomment-1576693005, I have been running the wrong tests. I'll have another look and keep you posted.

**Review by grebaldi — APPROVED** [review]
I will move those test cases regarding parentheses into the Parser realm in the scope of https://github.com/PackageFactory/PackageFactory.ComponentEngine/pull/31, once it's rebased on this.

**Comment by mhsdesign** [comment]
Thanks for taking care @grebaldi and sorry for not responding ^^ its hard to have so many side projects xD

File: src/Parser/Ast/ExpressionNode.php [hunk 6]
```diff
@@ -122,8 +123,14 @@ public static function fromTokens(\Iterator $tokens, Precedence $precedence = Pr
...
 
         Scanner::skipSpaceAndComments($tokens);
+        if (Scanner::isEnd($tokens) || $precedence->mustStopAt(Scanner::type($tokens))) {
+            return new self(
+                root: $root
+            );
+        }
 
         while (!Scanner::isEnd($tokens) && !$precedence->mustStopAt(Scanner::type($tokens))) {
+            Scanner::skipSpaceAndComments($tokens);
```
Comment (mhsdesign) [inline src/Parser/Ast/ExpressionNode.php:133]: The additions that are here are not necessary. The condition is also handled by the "while" loop and later comes the return. 

I think i messed stuff up because the partial fix belongs to https://github.com/PackageFactory/PackageFactory.ComponentEngine/pull/10

either way, it works as it is, just not super clean.
